altair/vegalite/v5/schema/channels.py

```diff
@@ -38,9 +38,12 @@ def to_dict(self, validate=True, ignore=(), context=None):
                 parsed.pop('type', None)
             elif not (type_in_shorthand or type_defined_explicitly):
                 if isinstance(context.get('data', None), pd.DataFrame):
-                    raise ValueError("{} encoding field is specified without a type; "
-                                     "the type cannot be inferred because it does not "
-                                     "match any column in the data.".format(shorthand))
+                    raise ValueError(
+                        'Unable to determine data type for the field "{}";'
+                        " verify that the field name is not misspelled."
+                        " If you are referencing a field from a transform,"
```

Comment:
But this phrasing is incomplete. Since datasets defined as url/json/csv (not pd.DataFrame) requires a type definition too.

Reply:
Those data sources are all handled on line 48 below. The error message here is only if we know that the data is a pandas data frame as per the check on line 40.

Reply:
But that error message is also incomplete, since those fields can also be misspelled instead of the missing type definition.

Reply:
No, that is inaccurate what I say, since that becomes https://github.com/altair-viz/altair/issues/2825..

Reply:
Actually they can't. that's what #2825 is about. If the type is correct, it doesn't matter if those fields are misspelled, the chart will still appear (but be empty). This error is only raised if there is no type detected.